controllers/dcgmexporter_controller.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package controllers contains the main controller, where the reconciliation starts.
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/manifests/dcgmexporter"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/apis/v1alpha1"
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/config"
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/manifests"
+	collectorStatus "github.com/aws/amazon-cloudwatch-agent-operator/internal/status/collector"
+)
+
+// DcgmExporterReconciler reconciles a DcgmExporter object.
+type DcgmExporterReconciler struct {
+	client.Client
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+	log      logr.Logger
+	config   config.Config
+}
+
+func (r *DcgmExporterReconciler) getParams(instance v1alpha1.DcgmExporter) manifests.Params {
+	return manifests.Params{
+		Config:   r.config,
+		Client:   r.Client,
+		DcgmExp:  instance,
+		Log:      r.log,
+		Scheme:   r.scheme,
+		Recorder: r.recorder,
+	}
+}
+
+// NewReconciler creates a new reconciler for DcgmExporter objects.
+func NewDcgmExporterReconciler(p Params) *DcgmExporterReconciler {
+	r := &DcgmExporterReconciler{
+		Client:   p.Client,
+		log:      p.Log,
+		scheme:   p.Scheme,
+		config:   p.Config,
+		recorder: p.Recorder,
+	}
+	return r
+}
+
+// +kubebuilder:rbac:groups="",resources=pods;configmaps;services;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+// +kubebuilder:rbac:groups=apps,resources=daemonsets;deployments;statefulsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=autoscaling,resources=horizontalpodautoscalers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=policy,resources=poddisruptionbudgets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;create;update
+// +kubebuilder:rbac:groups=monitoring.coreos.com,resources=servicemonitors;podmonitors,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=networking.k8s.io,resources=ingresses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=route.openshift.io,resources=routes;routes/custom-host,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cloudwatch.aws.amazon.com,resources=dcgmexporters,verbs=get;list;watch;update;patch
```

Comment:
what are all of these roles above cloudwatch.aws.amazon.com? Do we need those?

Reply:
It's for the controller to monitor dcgm exporter custom resources